src/jsoncharutils.h

```diff
@@ -324,10 +324,31 @@ struct value128 {
   uint64_t high;
 };
 
+#if defined(_MSC_VER) && !defined(_M_X64) // _umul128 for x86, arm, arm64
+#if defined(_M_ARM)
+static inline uint64_t __emulu(uint32_t x, uint32_t y) {
+  return x * (uint64_t)y;
+}
+#endif
+static inline uint64_t _umul128(uint64_t ab, uint64_t cd, uint64_t *hi) {
+  uint64_t ad = __emulu((uint32_t)(ab >> 32), (uint32_t)cd);
+  uint64_t bd = __emulu((uint32_t)ab, (uint32_t)cd);
+  uint64_t adbc = ad + __emulu((uint32_t)ab, (uint32_t)(cd >> 32));
+  uint64_t adbc_carry = !!(adbc < ad);
+  uint64_t lo = bd + (adbc << 32);
+  *hi = __emulu((uint32_t)(ab >> 32), (uint32_t)(cd >> 32)) + (adbc >> 32) + (adbc_carry << 32) + !!(lo < bd);
+  return lo;
+}
+#if defined(_M_ARM64)
+uint64_t umul128(uint64_t ab, uint64_t cd, uint64_t* hi) { // non-inline extern version for arm64
+  return _umul128(ab, cd, hi);
```

Comment:
This would only be acceptable if Visual Studio has no native _umul128. It is maybe not the case, but you can get the high 64-bits of a multiplication using a single instruction with the __umulh intrinsic, see...

 https://stackoverflow.com/questions/28868367/getting-the-high-part-of-64-bit-integer-multiplication

Reply:
So I think the proper approach under ARM 64 bits is...
```C++
hi = __umulh(ab, cd);
return ab * cd;
```